[mac_move] [mock dualtor] add wait time after swss restart

maunnikr-cisco opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
In test_orchagent_mac_move.py, a new neighbor learnt on an active port and we get the crm facts from DUT before and after sending packets to this new neighbor. crm polling interval is set to 1 at the start of the testcase and the timeout given in TC for getting the crm facts from DUT was 11 secs. Its taking more time than that populating the CRM facts on DUT causing the failure. Running the test_mac_move testcase was causing two types of errors :

- crm_facts was timing out
- even if crm_facts did not time out, the expected packet was not found after tcpdump

Summary:

Adds wait time of 100s after swss restart - allows time for updating of CRM, and for the application and updating of configs. 
Ensures that the mac_move testcase is sending packets to the right kind of ports (Active or Standby) in the particular part of the testcase, by only selecting from a list of PTF ports connecting to Active portchannels when trying to send to Active ports, and switching the configuration from Active to Standby when sending to Standby ports, so our expectations for the testcase are consistent with the results. The CRM update was passing earlier, but is failing now because of the skipped wait time owing to [This PR](https://github.com/sonic-net/sonic-buildimage/pull/15741) which skips the tunneling wait time of 90s for mock dualtor cases. 


### Type of change
- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205

### Approach
#### What is the motivation for this PR?
Ensuring that the expected test_orchagent_mac_move.py::test_mac_move passes, and is able to detect the expected packets

#### How did you do it?
- Added a wait time of 100s after swss restart - this compensates for the 90s of tunnel time that is skipped for Mock Dualtor, allowing time for CRM to be updated and for configs to be updated. Earlier, on skipping this time - CRM was not updated by the time we call get_crm_facts, which is why the function was timing out, and additionally, the first part of the testcase where the packets are sent to an Active port as a way to introduce a neighbor, was not happening since we were not allowing enough time for the configurations of the ports to be applied and for the port we were sending the packet on the to be Active in configuration. Therefore, we were sending the packets to a Standby PTF port, and expecting to see collect them at the PTF interface as we would in the case of an Active PTF port. 
- Replaced the function that provided the list of interfaces to pick randomly from with one that would give only the list of DUT PortChannels associated with Active PTF ports so that we were ensuring that we were testing with Active ports. 
- Replaced the while loop in get_crm_facts with a wait_until statement - compressed the code and made it more readable

#### How did you verify/test it?
Ran the command ./run_tests.sh -n t0-64-2 -l debug -u -c dualtor/test_orchagent_mac_move.py::test_mac_move -e -s  -p /run_logs/mac_move_malavika_24_8_final in the sonic-mgmt suite (/data/tests)

Logs in /run_logs/mac_move_malavika_24_8_final
<img width="1723" alt="image" src="https://github.com/sonic-net/sonic-mgmt/assets/142842429/de91b094-544b-477e-b314-0149cb8e577d">

<img width="1727" alt="image" src="https://github.com/sonic-net/sonic-mgmt/assets/142842429/0a048ac0-dd31-4702-92a1-4cc37dfa2088">

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation

maunnikr-cisco committed:
Add wait time after swss restart

Changing dual_tor_mock to add a wait time after swss restart, for the update of configs, and to provide buffer time for CRM update.

maunnikr-cisco committed:
get list of only active PTF ports

change the function that returns list of all PTF ports, to the one that returns the list of the DUT portchannels connected to Active PTF ports to ensure we are sending packets to only Active ports

maunnikr-cisco committed:
compressing while loop using wait_until

replacing while loop in get_crm_facts with a wait_until function

maunnikr-cisco committed:
Update test_orchagent_mac_move.py

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Failed<br/>- hook id: trailing-whitespace<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/dualtor/test_orchagent_mac_move.py<br/><br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/devices/sonic.py:1820:25: F823 local variable 'time_total' defined in enclosing scope on line 1857 referenced before assignment<br/>tests/common/devices/sonic.py:1820:25: F841 local variable 'time_total' is assigned to but never used<br/>tests/dualtor/test_orchagent_mac_move.py:91:56: E261 at least two spaces before inline comment<br/>tests/dualtor/test_orchagent_mac_move.py:91:121: E501 line too long (125 > 120 characters)<br/>tests/dualtor/test_orchagent_mac_move.py:92:67: E261 at least two spaces before inline comment<br/>...<br/>[truncated extra lines, please run pre-commit locally to view full check results]<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

maunnikr-cisco committed:
Merge branch 'sonic-net:master' into add-timeout

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Failed<br/>- hook id: trailing-whitespace<br/>- exit code: 1<br/>- files were modified by this hook<br/><br/>Fixing tests/dualtor/test_orchagent_mac_move.py<br/><br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/devices/sonic.py:1820:25: F823 local variable 'time_total' defined in enclosing scope on line 1857 referenced before assignment<br/>tests/common/devices/sonic.py:1820:25: F841 local variable 'time_total' is assigned to but never used<br/>tests/dualtor/test_orchagent_mac_move.py:91:56: E261 at least two spaces before inline comment<br/>tests/dualtor/test_orchagent_mac_move.py:91:121: E501 line too long (125 > 120 characters)<br/>tests/dualtor/test_orchagent_mac_move.py:92:67: E261 at least two spaces before inline comment<br/>...<br/>[truncated extra lines, please run pre-commit locally to view full check results]<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

maunnikr-cisco committed:
Update test_orchagent_mac_move.py

maunnikr-cisco committed:
Update sonic.py

maunnikr-cisco committed:
Fixing some syntax errors, and formatting issues

fixing variable scope issue in get_crm_facts, and re-formatting the code to match the standards

Aravind-Subbaroyan commented:
@lolyu - could you review this PR?

lolyu commented on tests/dualtor/test_orchagent_mac_move.py:
```diff
@@ -87,8 +87,14 @@ def test_mac_move(
+    # randomly select a portchannel from the list
+    portchannel = portchannels[random.choice(portchannels.keys())]
+    # randomly choose an active PTF port from the selected portchannel
+    index_n = random.choice(list(range(len(portchannel))))
```
```suggestion
    ptf_t1_intf_index = random.choice(portchannel)
```

maunnikr-cisco replied:
Have committed corresponding changes.

lolyu commented on tests/common/dualtor/dual_tor_mock.py:
```diff
@@ -350,6 +350,8 @@ def apply_peer_switch_table_to_dut(cleanup_mocked_configs, rand_selected_dut, mo
         dut.shell('systemctl reset-failed swss; systemctl restart swss')
         wait_critical_processes(dut)
+        logger.info("Wait for 100 seconds for complete update of configs after swss restart")
+        time.sleep(100)
```
could you please elaborate why do we need wait for 100s?

lolyu replied:
Can we improve this with a `wait_until` that has a check to see if the tunnel intf is up and ready?

maunnikr-cisco replied:
Reason for 100s wait : 
[This PR](https://github.com/sonic-net/sonic-buildimage/pull/15741) is skipping the wait for tunnel for mock dualtor - 
the reason for this is described in the PR, but basically since we are skipping wait_for_tunnel on mocked dualtor, we are skipping the 90s wait time that was earlier meant for tunneling - however previously, during this 90s wait time application and updating of configs after swss restart was completed, and CRM was also updated. Therefore, when we skip this, configs are not applied, and CRM is not updated, and therefore this 100s wait time after swss restart compensates for the same.

maunnikr-cisco committed:
Simplifying ptf port selection

optimizing code to make it shorter and simpler

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/dualtor/test_orchagent_mac_move.py:96:51: E261 at least two spaces before inline comment<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

maunnikr-cisco committed:
Fixing format issue 

Adding extra space for inline comment

maunnikr-cisco committed:
Merge branch 'master' into add-timeout

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/devices/sonic.py:1918:9: F821 undefined name 'wait_until'<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

maunnikr-cisco committed:
Merge branch 'sonic-net:master' into add-timeout

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/common/devices/sonic.py:1918:9: F821 undefined name 'wait_until'<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

maunnikr-cisco commented:
[test_rochagent.log](https://github.com/sonic-net/sonic-mgmt/files/14542192/test_rochagent.log)

Logs from successful run

maunnikr-cisco committed:
import wait_until

maunnikr-cisco committed:
Update test_orchagent_mac_move.py

maunnikr-cisco committed:
Merge pull request #3 from maunnikr-cisco/maunnikr-cisco-patch-1

import wait_until

mssonicbld commented:
The [pre-commit](http://pre-commit.com/) check detected issues in the files touched by this pull request.
The pre-commit check is a mandatory check, please fix detected issues.

Detailed pre-commit check results:
<samp>trim trailing whitespace.................................................Passed<br/>fix end of files.........................................................Passed<br/>check yaml...........................................(no files to check)Skipped<br/>check for added large files..............................................Passed<br/>check python ast.........................................................Passed<br/>flake8...................................................................Failed<br/>- hook id: flake8<br/>- exit code: 1<br/><br/>tests/dualtor/test_orchagent_mac_move.py:16:1: F401 'tests.common.utilities.wait_until' imported but unused<br/><br/>flake8...............................................(no files to check)Skipped<br/>check conditional mark sort..........................(no files to check)Skipped<br/></samp>

To run the pre-commit checks locally, you can follow below steps:
1. Ensure that default python is python3. In sonic-mgmt docker container, default python is python2. You can run
   the check by activating the python3 virtual environment in sonic-mgmt docker container or outside of sonic-mgmt
   docker container.
2. Ensure that the `pre-commit` package is installed:
```
sudo pip install pre-commit
```
3. Go to repository root folder
4. Install the pre-commit hooks:
```
pre-commit install
```
5. Use pre-commit to check staged file:
```
pre-commit
```
6. Alternatively, you can check committed files using:
```
pre-commit run --from-ref <commit_id> --to-ref <commit_id>
```

maunnikr-cisco committed:
Remove unnecessary wait_until import

kevinskwang commented:
/azp run

maunnikr-cisco committed:
Merge branch 'master' into add-timeout

kevinskwang commented on tests/common/devices/sonic.py:
```diff
@@ -16,8 +16,10 @@
 
 from tests.common.devices.base import AnsibleHostBase
 from tests.common.devices.constants import ACL_COUNTERS_UPDATE_INTERVAL_IN_SEC
+
```
please remove the redundant line

maunnikr-cisco replied:
Done.

kevinskwang commented on tests/dualtor/test_orchagent_mac_move.py:
```diff
@@ -88,8 +88,13 @@ def test_mac_move(
     tor = rand_selected_dut
-    ptf_t1_intf = random.choice(get_t1_ptf_ports(tor, tbinfo))
-    ptf_t1_intf_index = int(ptf_t1_intf.strip("eth"))
+    portchannels = None
```
what's the purpose of this code section? I didn't see any description in the commit message

maunnikr-cisco replied:
HI Kevin, the purpose of this code is to select an active PTF port, since we were previously choosing randomly from all t1 ptf ports, and this was causing intermittent error, since some ports would be down.

maunnikr-cisco committed:
Merge branch 'master' into add-timeout

mssonicbld commented:
/azp run

maunnikr-cisco committed:
Adding a config check after swss restart

mssonicbld commented:
/azp run

maunnikr-cisco committed:
Removing redundant space

mssonicbld commented:
/azp run

wsycqyz reviewed (dismissed):
lğtm

wsycqyz commented:
There is static analysis error. Please help fix it:

tests/common/devices/sonic.py:2078:9: F821 undefined name 'wait_until'
tests/dualtor/test_orchagent_mac_move.py:8:1: F401 'tests.common.dualtor.dual_tor_utils.get_t1_ptf_ports' imported but unused
tests/dualtor/test_orchagent_mac_move.py:93:20: F821 undefined name 'get_t1_active_ptf_ports'

maunnikr-cisco committed:
Fixing some import errors

mssonicbld commented:
/azp run

maunnikr-cisco committed:
Fixing wait_until import

mssonicbld commented:
/azp run

maunnikr-cisco committed:
Update test_orchagent_mac_move.py

maunnikr-cisco committed:
Adding wait_until import in the right place

maunnikr-cisco committed:
Merge pull request #4 from maunnikr-cisco/maunnikr-cisco-patch-1

Change wait_until import location

maunnikr-cisco commented:
/azp run

maunnikr-cisco committed:
Update test_orchagent_mac_move.py

mssonicbld commented:
/azp run
